presto-main/src/main/java/com/facebook/presto/type/Re2JRegexp.java

```diff
@@ -0,0 +1,160 @@
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.facebook.presto.type;
+
+import com.facebook.presto.spi.PrestoException;
+import com.facebook.presto.spi.block.Block;
+import com.facebook.presto.spi.block.BlockBuilder;
+import com.facebook.presto.spi.block.BlockBuilderStatus;
+import com.google.re2j.Matcher;
+import com.google.re2j.Options;
+import com.google.re2j.Pattern;
+import io.airlift.log.Logger;
+import io.airlift.slice.Slice;
+
+import static com.facebook.presto.spi.StandardErrorCode.INVALID_FUNCTION_ARGUMENT;
+import static com.facebook.presto.spi.type.VarcharType.VARCHAR;
+import static com.google.common.base.Preconditions.checkState;
+import static com.google.common.primitives.Ints.checkedCast;
+import static com.google.re2j.Options.Algorithm.DFA_FALLBACK_TO_NFA;
+import static java.lang.String.format;
+
+public final class Re2JRegexp
+{
+    private static final Logger log = Logger.get(Re2JRegexp.class);
+
+    private static final java.util.regex.Pattern DOT_STAR_PREFIX_SUFFIX_PATTERN = java.util.regex.Pattern.compile("^(\\.\\*\\??)?(.*?)(\\.\\*\\??)?$");
+    private static final int CORE_PATTERN_INDEX = 2;
+
+    public final int dfaStatesLimit;
+    public final int dfaRetries;
+
+    public final Pattern re2jPattern;
+    public final Pattern re2jPatternWithoutDotStartPrefixSuffix;
+
+    public Re2JRegexp(int dfaStatesLimit, int dfaRetries, Slice pattern)
+    {
+        this.dfaStatesLimit = dfaStatesLimit;
+        this.dfaRetries = dfaRetries;
+
+        Options options = Options.builder()
+                .setAlgorithm(DFA_FALLBACK_TO_NFA)
+                .setMaximumNumberOfDFAStates(dfaStatesLimit)
+                .setNumberOfDFARetries(dfaRetries)
+                .setEventsListener(new RE2JEventsListener())
+                .build();
+
+        String patternString = pattern.toStringUtf8();
+        re2jPattern = Pattern.compile(patternString, options);
+
+        // Remove .* prefix and suffix. This will give performance boost when using RE2J Pattern.find() function.
+        java.util.regex.Matcher dotStarPrefixSuffixMatcher = DOT_STAR_PREFIX_SUFFIX_PATTERN.matcher(patternString);
+        checkState(dotStarPrefixSuffixMatcher.matches());
+        String patternStringWithoutDotStartPrefixSuffix = dotStarPrefixSuffixMatcher.group(CORE_PATTERN_INDEX);
+
+        if (!patternStringWithoutDotStartPrefixSuffix.equals(patternString)) {
+            re2jPatternWithoutDotStartPrefixSuffix = Pattern.compile(patternStringWithoutDotStartPrefixSuffix, options);
+        }
+        else {
+            re2jPatternWithoutDotStartPrefixSuffix = re2jPattern;
+        }
+    }
+
+    public boolean matches(Slice source)
+    {
+        return re2jPatternWithoutDotStartPrefixSuffix.find(source);
```

Comment:
I have run basic C++ RE2 benchmark to confirm that it works the same:
```
Search_Easy0_CachedRE2/8	10000000	       191 ns/op	  41.74 MB/s
Search_Easy0_CachedRE2/16	10000000	       198 ns/op	  80.77 MB/s
Search_Easy0_CachedRE2/32	10000000	       200 ns/op	 159.77 MB/s
Search_Easy0_CachedRE2/64	10000000	       211 ns/op	 302.34 MB/s
Search_Easy0_CachedRE2/128	10000000	       213 ns/op	 599.13 MB/s
Search_Easy0_CachedRE2/256	10000000	       230 ns/op	1109.17 MB/s
Search_Easy0_CachedRE2/512	10000000	       283 ns/op	1806.49 MB/s
Search_Easy0_CachedRE2/1K	 5000000	       385 ns/op	2657.43 MB/s
Search_Easy0_CachedRE2/2K	 5000000	       479 ns/op	4269.04 MB/s
Search_Easy0_CachedRE2/4K	 5000000	       741 ns/op	5520.85 MB/s
Search_Easy0_CachedRE2/8K	 1000000	      1453 ns/op	5637.17 MB/s
Search_Easy0_CachedRE2/16K	  500000	      2775 ns/op	5903.13 MB/s
Search_Easy0_CachedRE2/32K	  500000	      5672 ns/op	5776.17 MB/s
Search_Easy0_CachedRE2/64K	  200000	     12422 ns/op	5275.48 MB/s
Search_Easy0_CachedRE2/128K	  100000	     28655 ns/op	4574.12 MB/s
Search_Easy0_CachedRE2/256K	   50000	     63872 ns/op	4104.17 MB/s
Search_Easy0_CachedRE2/512K	   10000	    138427 ns/op	3787.46 MB/s
Search_Easy0_CachedRE2/1M	   10000	    285923 ns/op	3667.33 MB/s
Search_Easy0_CachedRE2/2M	    5000	    561748 ns/op	3733.26 MB/s
Search_Easy0_CachedRE2/4M	    2000	   1081669 ns/op	3877.62 MB/s
Search_Easy0_CachedRE2/8M	    1000	   2351800 ns/op	3566.89 MB/s
Search_Easy0_CachedRE2/16M	     500	   4837912 ns/op	3467.86 MB/s
Search_Hard_CachedRE2/8	10000000	       216 ns/op	  36.89 MB/s
Search_Hard_CachedRE2/16	10000000	       248 ns/op	  64.30 MB/s
Search_Hard_CachedRE2/32	 5000000	       304 ns/op	 105.06 MB/s
Search_Hard_CachedRE2/64	 5000000	       402 ns/op	 158.96 MB/s
Search_Hard_CachedRE2/128	 5000000	       600 ns/op	 213.04 MB/s
Search_Hard_CachedRE2/256	 2000000	       962 ns/op	 266.09 MB/s
Search_Hard_CachedRE2/512	 1000000	      1686 ns/op	 303.62 MB/s
Search_Hard_CachedRE2/1K	 1000000	      3266 ns/op	 313.51 MB/s
Search_Hard_CachedRE2/2K	  500000	      6585 ns/op	 310.97 MB/s
Search_Hard_CachedRE2/4K	  200000	     12690 ns/op	 322.76 MB/s
Search_Hard_CachedRE2/8K	  100000	     24019 ns/op	 341.05 MB/s
Search_Hard_CachedRE2/16K	   50000	     49293 ns/op	 332.37 MB/s
Search_Hard_CachedRE2/32K	   10000	    100712 ns/op	 325.36 MB/s
Search_Hard_CachedRE2/64K	   10000	    198442 ns/op	 330.25 MB/s
Search_Hard_CachedRE2/128K	    5000	    408571 ns/op	 320.81 MB/s
Search_Hard_CachedRE2/256K	    5000	    794448 ns/op	 329.97 MB/s
Search_Hard_CachedRE2/512K	    1000	   1538259 ns/op	 340.83 MB/s
Search_Hard_CachedRE2/1M	     500	   3212420 ns/op	 326.41 MB/s
Search_Hard_CachedRE2/2M	     500	   6503830 ns/op	 322.45 MB/s
Search_Hard_CachedRE2/4M	     100	  12913210 ns/op	 324.81 MB/s
Search_Hard_CachedRE2/8M	      50	  26015640 ns/op	 322.44 MB/s
Search_Hard_CachedRE2/16M	      50	  50779500 ns/op	 330.39 MB/s
```
Easy pattern is:
`ABCDEFGHIJKLMNOPQRSTUVWXYZ$`
Hard pattern is:
`.*ABCDEFGHIJKLMNOPQRSTUVWXYZ$`